runtime/eventlog.c

```diff
@@ -0,0 +1,269 @@
+#define CAML_INTERNALS
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <inttypes.h>
+#include "caml/alloc.h"
+#include "caml/eventlog.h"
+#include "caml/misc.h"
+#include "caml/osdeps.h"
+
+#define CTF_MAGIC 0xc1fc1fc1
+#define CAML_TRACE_VERSION 0x1
+
+struct ctf_stream_header {
+  uint32_t magic;
+  uint16_t caml_trace_version;
+  uint16_t stream_id;
+};
+
+static struct ctf_stream_header header = {
+  CTF_MAGIC,
+  CAML_TRACE_VERSION,
+  0
+};
+
+#pragma pack(1)
+struct ctf_event_header {
+  uint64_t timestamp;
+  uint32_t pid;
+  uint32_t id;
+};
+
+static uintnat alloc_buckets [20] =
+  {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+
+
+struct event {
+  struct ctf_event_header header;
+  uint8_t  phase; /* for GC events */
+  uint8_t  counter_kind; /* misc counter name */
+  uint8_t  alloc_bucket; /* for alloc counters */
+  uint32_t count; /* for misc counters */
+};
+
+static FILE* output;
+static uintnat eventlog_startup_timestamp = 0;
+static uint32_t eventlog_startup_pid = 0;
+static uintnat eventlog_paused = 0;
+
+uintnat caml_eventlog_enabled = 0;
+
+#define EVENT_BUF_SIZE 4096
+struct event_buffer {
+  uintnat ev_generated;
+  struct event events[EVENT_BUF_SIZE];
+};
+
+static struct event_buffer* evbuf;
+
+void setup_evbuf()
+{
+  CAMLassert(!evbuf);
+  evbuf = caml_stat_alloc_noexc(sizeof(*evbuf));
+
+  if (evbuf == NULL)
+    caml_fatal_error("eventlog: could not allocate event buffer");
+
+  evbuf->ev_generated = 0;
+}
+
+void setup_eventlog_file()
+{
+  char_os *filename = NULL;
+  char_os *eventlog_filename;
+  value tmp;
+
+  eventlog_filename = caml_secure_getenv(T("OCAML_EVENTLOG_FILE"));
+  if (eventlog_filename) {
+    char *input =  caml_stat_strdup_of_os(eventlog_filename);
+
+    if (input != NULL) {
+      tmp = caml_alloc_sprintf("%s.%d.eventlog",
+                               input, eventlog_startup_pid);
+      caml_stat_free(input);
+      filename = caml_stat_strdup_to_os(String_val(tmp));
+    }
+
+  } else {
+    tmp = caml_alloc_sprintf("caml-eventlog-%d", eventlog_startup_pid);
+    filename = caml_stat_strdup_to_os(String_val(tmp));
+  }
+
+  if (filename) {
+    output = fopen_os(filename, T("wb"));
+    caml_stat_free(filename);
+  }
+
+  if (output) {
+    fwrite(&header, sizeof(struct ctf_stream_header), 1, output);
+    fflush(output);
+  } else {
+    caml_eventlog_enabled = 0;
```

Comment:
I think the API would be slightly nicer if this function returned a boolean (`0` when it failed to setup the event log, `1` if the output is correctly initialized), and the caller was in charge of disabling event-logging in the failure case.